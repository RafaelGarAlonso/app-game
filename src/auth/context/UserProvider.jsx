import { UserContext } from './UserContext';
import { INIT_USER, INIT_USERS_TABLE } from '../../constants/users'

export const UserProvider = ({ children }) => {

  const getUser = () => {
    const localStorageData = localStorage.getItem('currentUser');
    return localStorageData ? JSON.parse(localStorageData) : INIT_USER;
  };

  const setNewUserSession = ({ userName }) => {
    const listOfUsers = getListOfUsers();

    for (const [index, user] of listOfUsers.entries()) {
      if (user.userName === userName) {
        const { userName, points } = listOfUsers[index];
        localStorage.setItem('currentUser', JSON.stringify({ userName, points }));
        return false;
      }
    };

    localStorage.setItem('currentUser', JSON.stringify({ userName, points: 0 }));
    return true;
  };

  const getAppLoaded = () => {
    return sessionStorage.getItem('appLoaded');
  };

  const setAppLoaded = (loaded) => {
    sessionStorage.setItem('appLoaded', loaded);
  };

  const userEarnPoints = () => {
    const user = getUser();
    localStorage.setItem('currentUser', JSON.stringify({ userName: user.userName, points: user.points + 1 }));
  };

  const getListOfUsers = () => {
    const listOfUsers = localStorage.getItem('listOfUsers');
    return listOfUsers ? JSON.parse(listOfUsers) : null;
  };

  const setNewListOfUsers = (listOfUsers = INIT_USERS_TABLE) => {
    localStorage.setItem('listOfUsers', JSON.stringify(listOfUsers));
  };

  const setPointsToListUsers = (currentUser) => {
    const listOfUsers = getListOfUsers();

    for (let [index, user] of listOfUsers.entries()) {
      if (user.userName === currentUser.userName) {
        listOfUsers[index].points = currentUser.points;
        break;
      }
    };

    setNewListOfUsers(listOfUsers);
  };

  return (
    <UserContext.Provider value={{
      setNewUserSession,
      getUser,
      setAppLoaded,
      getAppLoaded,
      userEarnPoints,
      getListOfUsers,
      setNewListOfUsers,
      setPointsToListUsers
    }}>
      {children}
    </UserContext.Provider>
  )
}