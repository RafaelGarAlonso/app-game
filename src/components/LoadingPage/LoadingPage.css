.container-loading {
    width: 100%;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

.loading-image {
    display: flex;
    width: 50%;
    justify-content:center;
    margin: 50px auto 60px auto;
    border: 1px dotted #747474;
    padding: 0;
    -webkit-animation:spin 4s linear infinite;
    -moz-animation:spin 4s linear infinite;
    animation:spin 4s linear infinite;
}

.title {
    display: flex;
    justify-content: center;
    width: 100%;
    text-decoration: underline;
    font-weight: 700;
    text-align: center;
}

@-moz-keyframes spin { 
    100% { -moz-transform: rotate(360deg); } 
}
@-webkit-keyframes spin { 
    100% { -webkit-transform: rotate(360deg); } 
}
@keyframes spin { 
    100% { 
        -webkit-transform: rotate(360deg); 
        transform:rotate(360deg); 
    } 
}