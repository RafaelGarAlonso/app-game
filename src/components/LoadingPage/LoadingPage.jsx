import { AnimatedProgressBar } from '../ProgressBar/ProgressBar';
import Image from 'react-bootstrap/Image';
import { useNavigate } from 'react-router-dom';
import { useContext, useEffect } from 'react';
import { UserContext } from '../../auth/context/UserContext';
import "./LoadingPage.css"

export const LoadingPage = () => {
    const navigate = useNavigate();
    const { setAppLoaded } = useContext(UserContext);

    useEffect(() => {
        setAppLoaded(false);
    }, [setAppLoaded]);

    const redirectToHome = () => {
        setAppLoaded(true);
        setTimeout(() => {
            navigate('/home', { replace: true });
        }, 1000);
    }

    return (
        <div className="container-loading">
            <div className="row p-4">
                <h1 className="title">Welcome to <br />Â¡Rock Paper Scissors Shoot!</h1>
                <Image className="loading-image" src="assets/loading-background.png" roundedCircle />
                <AnimatedProgressBar onComplete={redirectToHome} timeToComplete={25} />
            </div>
        </div>
    )
}