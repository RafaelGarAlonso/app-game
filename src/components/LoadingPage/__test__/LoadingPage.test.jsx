import { render, act } from '@testing-library/react';
import { LoadingPage } from '../LoadingPage';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../../auth/context/UserProvider';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockedUsedNavigate,
}));

describe('Test <LoadingPage />', () => {

    beforeEach(() => {
        jest.useFakeTimers();
        jest.spyOn(global, 'setInterval');
        jest.spyOn(global, 'clearInterval');
    });

    const renderTemplate = () => {
        return render(
            <UserProvider>
                <MemoryRouter>
                    <LoadingPage />
                </MemoryRouter>
            </UserProvider>
        );
    }

    test('Check snapshot component', () => {
        const { container } = renderTemplate();
        expect(container).toMatchSnapshot();
    });

    // TODO: Fix Callback complete
    test('It should redirect to /home when progress bar finished', async () => {
        const delay = 2500;

        act(() => {
            jest.advanceTimersByTime(delay);
        });
    });

});