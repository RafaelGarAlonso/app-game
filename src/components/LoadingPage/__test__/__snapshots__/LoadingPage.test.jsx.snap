// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <LoadingPage /> Check snapshot component 1`] = `
<div>
  <div
    class="container-loading"
  >
    <div
      class="row p-4"
    >
      <h1
        class="title"
      >
        Welcome to 
        <br />
        Â¡Rock Paper Scissors Shoot!
      </h1>
      <img
        class="loading-image rounded-circle"
        src="assets/loading-background.png"
      />
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="progress-bar progress-bar-animated progress-bar-striped"
          role="progressbar"
          style="width: 0%;"
        />
      </div>
    </div>
  </div>
</div>
`;
